cluster-autoscaler/simulator/delta_cluster_snapshot.go

```diff
@@ -0,0 +1,523 @@
+limitations under the License.
+*/
+
+package simulator
+
+import (
+	"fmt"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	schedulerlisters "k8s.io/kubernetes/pkg/scheduler/listers"
+	schedulernodeinfo "k8s.io/kubernetes/pkg/scheduler/nodeinfo"
+)
+
+// DeltaClusterSnapshot is implementation of ClusterSnapshot optimized for typical Cluster Autoscaler usage - (fork, add stuff, revert) repeated many times per loop.
+//
+// Complexity of some notable operations:
+//	fork - O(1)
+//	revert - O(1)
+//	commit - O(n)
+//	list all pods (no filtering) - O(n), cached
+//	list all pods (with filtering) - O(n)
+//	list node infos - O(n), cached
+//	get all nodes - O(n)
+//
+// Watch out for:
+//	node deletions, pod additions & deletions - invalidates cache of current snapshot
+//		(when forked affects delta, but not base.)
+//	pod affinity - causes scheduler framework to list pods with non-empty selector,
+//		so basic caching doesn't help.
+//
+type DeltaClusterSnapshot struct {
+	data *internalDeltaSnapshotData
+}
+
+type internalDeltaSnapshotDataNodeLister internalDeltaSnapshotData
+type internalDeltaSnapshotDataPodLister internalDeltaSnapshotData
+
+type internalDeltaSnapshotData struct {
+	baseData *internalDeltaSnapshotData
+
+	nodeInfoMap      map[string]*schedulernodeinfo.NodeInfo
+	deletedNodeInfos map[string]bool
+
+	nodeInfoList         []*schedulernodeinfo.NodeInfo
+	completeNodeInfoList []*schedulernodeinfo.NodeInfo
+	podList              []*apiv1.Pod
+	rawPods              [][]*apiv1.Pod
+	rawPodsTotal         int
+	havePodsWithAffinity []*schedulernodeinfo.NodeInfo
+}
+
+func (data *internalDeltaSnapshotData) GetNodeInfo(name string) (*schedulernodeinfo.NodeInfo, error) {
+	if data == nil {
+		return nil, fmt.Errorf("node not found")
+	}
+	if nodeInfo, found := data.nodeInfoMap[name]; found {
+		return nodeInfo, nil
+	}
+	if data.deletedNodeInfos[name] || data.baseData == nil {
+		return nil, fmt.Errorf("node not found")
+	}
+	return data.baseData.GetNodeInfo(name)
+}
+
+func (data *internalDeltaSnapshotData) GetNodeInfoList() []*schedulernodeinfo.NodeInfo {
+	if data == nil {
+		return nil
+	}
+	if data.completeNodeInfoList == nil {
+		data.buildNodeInfoList()
+	}
+	return data.completeNodeInfoList
+}
+
+// buildNodeInfoList contains costly copying throughout the struct chain. Use wisely.
+func (data *internalDeltaSnapshotData) buildNodeInfoList() {
+	if data.nodeInfoList == nil {
+		data.nodeInfoList = make([]*schedulernodeinfo.NodeInfo, len(data.nodeInfoMap), len(data.nodeInfoMap)+100)
+		i := 0
+		for _, node := range data.nodeInfoMap {
+			data.nodeInfoList[i] = node
+			i++
+		}
+	}
+	if data.baseData == nil {
+		data.completeNodeInfoList = data.nodeInfoList
+		return
+	}
+	baseList := data.baseData.GetNodeInfoList()
+	totalLen := len(baseList) + len(data.nodeInfoList)
+	data.completeNodeInfoList = make([]*schedulernodeinfo.NodeInfo, totalLen, totalLen+100)
+	if len(data.deletedNodeInfos) > 0 {
```

Comment:
This index based list building looks error prone. 
Why not just use `append` in both if branches
 - one by one in the loop in first if branch
 - and append(cNIL, baseList...) in else branch

Reply:
Changed, PTAL.